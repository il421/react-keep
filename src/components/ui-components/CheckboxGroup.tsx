import React from "react";

import { JustifyContent } from "../../common";
import "../../styles/ui-components/_login-button.scss";
import { FlexBox } from "./FlexBox";

export interface Option {
  id: string;
  value: string;
}
export interface CheckboxGroupProps {
  options: Option[];
  selectedKeys: string[];
  onChange: (selectedKeys: string[]) => void;
  classNames?: {
    option?: string;
    checkbox?: string;
    label?: string;
  };
}

export class CheckboxGroup extends React.PureComponent<CheckboxGroupProps> {
  private handleOnChange = (selectedKey: string) => {
    let selectedKeys: string[] = this.props.selectedKeys || [];

    const isAlreadySelected = selectedKeys.includes(selectedKey);
    if (isAlreadySelected) {
      selectedKeys = selectedKeys.filter(x => x !== selectedKey);
    } else {
      selectedKeys = [...selectedKeys, selectedKey];
    }
    if (this.props.onChange) {
      this.props.onChange(selectedKeys);
    }
  };

  render() {
    const { options, classNames, selectedKeys } = this.props;

    return (
      <>
        {options.map((o: Option) => {
          const checked = selectedKeys.includes(o.id);

          return (
            <FlexBox
              key={o.id}
              justifyContent={JustifyContent.start}
              className={classNames?.option}
            >
              <div className={classNames?.checkbox}>
                <input
                  ref={o.id}
                  id={o.id}
                  value={o.id}
                  type="checkbox"
                  defaultChecked={checked}
                />
                <label
                  htmlFor={o.id}
                  onClick={() => this.handleOnChange(o.id)}
                />
              </div>

              <span className={classNames?.label}>{o.value}</span>
            </FlexBox>
          );
        })}
      </>
    );
  }
}
