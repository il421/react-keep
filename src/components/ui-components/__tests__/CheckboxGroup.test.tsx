import { mount, ReactWrapper } from "enzyme";
import React from "react";

import { CheckboxGroup, CheckboxGroupProps } from "../CheckboxGroup";

let props: CheckboxGroupProps, wrapper: ReactWrapper<CheckboxGroupProps, {}>;

beforeEach(() => {
  props = {
    selectedKeys: ["id1"],
    options: [
      {
        id: "id1",
        value: "1"
      },
      {
        id: "id2",
        value: "2"
      }
    ],
    onChange: jest.fn()
  };
  wrapper = mount<CheckboxGroupProps>(<CheckboxGroup {...props} />);
});

test("should render CheckboxGroup, and contain input element", () => {
  expect(wrapper.debug()).toMatchSnapshot();
});

test("should select options correctly", () => {
  wrapper
    .find("#id2")
    .hostNodes()
    .simulate("change", { target: { value: true } });
  wrapper.find("label[htmlFor='id2']").hostNodes().simulate("click");
  expect(props.onChange).toHaveBeenLastCalledWith(["id1", "id2"]);
});
