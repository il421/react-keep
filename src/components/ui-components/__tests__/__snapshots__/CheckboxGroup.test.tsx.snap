// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CheckboxGroup, and contain input element 1`] = `
"<CheckboxGroup selectedKeys={{...}} options={{...}} onChange={[Function: mockConstructor]}>
  <FlexBox justifyContent=\\"flex-start\\" className={[undefined]}>
    <div style={{...}} className={[undefined]}>
      <div className={[undefined]}>
        <input id=\\"id1\\" value=\\"id1\\" type=\\"checkbox\\" defaultChecked={true} />
        <label htmlFor=\\"id1\\" onClick={[Function: onClick]} />
      </div>
      <span className={[undefined]}>
        1
      </span>
    </div>
  </FlexBox>
  <FlexBox justifyContent=\\"flex-start\\" className={[undefined]}>
    <div style={{...}} className={[undefined]}>
      <div className={[undefined]}>
        <input id=\\"id2\\" value=\\"id2\\" type=\\"checkbox\\" defaultChecked={false} />
        <label htmlFor=\\"id2\\" onClick={[Function: onClick]} />
      </div>
      <span className={[undefined]}>
        2
      </span>
    </div>
  </FlexBox>
</CheckboxGroup>"
`;
